### DISCUSSION TITLE
How to get display to show as rotated on VNC client/viewer?

### DISCUSSION THREAD

**Discussion by jegfish** [any1/wayvnc#128] [Q&A]
## Goal
Rotate the display so I can use the screen of the device my VNC client is running on in landscape mode, rather than the default portrait mode I am currently stuck with.

## Problem

My display shows as rotated on the main output, the virtual machine (VM) window. On the VNC client it does not show as rotated. However, it does seem to send the rotated resolution, as one VNC client ([VNSee](https://github.com/matteodelabre/vnsee)) reports the resolution as the transformed version; i.e. in swaywm the display's resolution is 1872x1404, and the display has a 90 degree rotation, VNSee reports the resolution as 1404x1872 (numbers flipped as a 90 degree rotation should do).

So while the resolution is "rotated", the windows and other things on screen are not. They still show as "right-side up", as if there was no rotation.

I've checked wayvnc's [FAQ](https://github.com/any1/wayvnc/blob/master/FAQ.md) and manual. I searched the manual (and read through as it is short and sweet) for "rotate", "rotat", "transform", but could not find anything. I did similar searches on the GitHub issues and discussions and couldn't find any solutions.

## My Setup

Host hardware: M1 ARM MacBook Air
VM software: [UTM for MacOS](https://mac.getutm.app/), which is a GUI wrapper around QEMU
VM guest OS: Fedora 35 aarch64

* Sway window manager display config:
```sh
$ swaymsg create_output
$ swaymsg 'output "HEADLESS-1" mode 1872x1404'
$ swaymsg 'output "HEADLESS-1" transform 90'
```
(default name of display created by `swaymsg create_output` is `HEADLESS-1`, assuming name not already taken)

* wayvnc command: `wayvnc --output=HEADLESS-1 --max-fps=10 10.11.99.2 5900`

### Further Details
I have the VM forwarding it's port 5900 to an open port on my host OS.

My ultimate goal is to get this to work on VNSee, but I also tried another VNC client to make sure it was not an issue with VNSee (I am not sure it is an issue with wayvnc, or an issue at all, it may be something I am doing or have not done). I tried TigerVNC both on the host OS and from within the VM guest. On the host OS it was clearly not rotated. In the guest it was a bit more confusing as I was the screen was rotated by 90 degrees :), but since the VNC client window was not rotated by another 90 degrees than the rest of the screen I believe it was also not rotated there.

**Comment by any1** [comment]
When you're sharing a physical rotated display, you want to see that display in the VNC client as you see it on your desk. That's why wayvnc reverses the rotation of the framebuffer.

**Reply by any1** [reply]
You could change this line https://github.com/any1/wayvnc/blob/master/src/main.c#L596 and set `output_transform = WL_OUTPUT_TRANSFORM_NORMAL` or any rotation that you want. You could even add a command line option to set an arbitrary output transform. PRs welcome. ;)

**Reply by jegfish** [reply]
Thank you, I managed to get it working. Once I got it working I realized the setup I was going for was not going to work out. Since it's not a use-case for me anymore, I've lost interest, but I appreciate that you welcome contributions to the project.

**Reply by captn3m0** [reply]
@jegfish I think I'm trying the same setup (Remarkable 2 with wayvnc + vnsee), and can't seem to get it working. It works perfectly fine in portrait, but any rotations on Sway side break it. I've tried lots of combinations of rotation + resolution configurations, but was wondering if you remember what config worked for you?

I've tried the below patch on wayvnc as well, but it didn't seem to help.

```
static void apply_output_transform(const struct wayvnc* self,
		struct wv_buffer* buffer, struct pixman_region16* damage)
{
	enum wl_output_transform output_transform, buffer_transform;
	output_transform = WL_OUTPUT_TRANSFORM_NORMAL;
```
